Starter site integration

adam-vessey opened the pull request:
**GitHub Issue**:
* https://github.com/Islandora/islandora-starter-site/issues/2
* https://github.com/Islandora/islandora_defaults/issues/74

* Other Relevant Links (Google Groups discussion, related pull requests,
 Release pull requests, etc.)
    * [Code Chill: Islandora Defaults](https://groups.google.com/g/islandora/c/uGzhTnW4TUI)
    * [Islandora Foundation contracting with DGI for "Modularizing Islandora" work](https://groups.google.com/g/islandora/c/vO1JAPO7GXM)

# What does this Pull Request do?

Gets the new "starter site" project installable via ansible.

# What's new?

Added the `starter` and `starter_dev` profiles, with minor reworks to allow some tasks to be suppressed.

* Does this change require documentation to be updated? Unknown.
* Does this change add any new dependencies? No/only if using the new "starter" profile(s).
* Does this change require any other modifications to be made to the repository
 (i.e. Regeneration activity, etc.)? No.
* Could this change impact execution of existing code? Potentially, with how some of the ansible state had to be manipulated... it seems some things were referring to the wrong state (such as the Matomo things defining the `set_search_api_config` value, and "Solr" stuff still referring to it). Also, if there's other 

# How should this be tested?

Smoketest and compare the behaviours of the `standard` and `demo` "profiles" from before and after the changes made here, they should behave the same as they did previously.

Test out the new `starter` and `starter_dev` "profiles". Should come up and look much the same, with the only difference presently being that `_dev` performs a clone of the project (which is only really evident when looking at how it the project exists as it was deployed on the filesystem).

# Additional Notes:

Nothing to mind...

# Interested parties
Tag (@ mention) interested parties or, if unsure, @Islandora-Devops/committers

adam-vessey committed:
Allow the installation separate installation of isladnora/openseadragon to be skipped.

adam-vessey committed:
Theoretically most of it?

... I think there's just one other section where there's a check for
things being explicitly "standard" that we should move over to a shared
variable?

adam-vessey committed:
More granular control over some drupal tasks.

adam-vessey committed:
Add the base starter_dev thing.

adam-vessey committed:
Allow matomo and fits to not be installed.

adam-vessey committed:
Fix new "when" specs, and deconflate Matomo and Solr configs.

adam-vessey committed:
Adjust/expand documentation of the "profiles".

adam-vessey commented on roles/internal/webserver-app/tasks/drupal.yml:
```diff
@@ -21,11 +21,53 @@
+  when: webserver_app_do_original_trusted_host_and_more
+
+- name: Trusted Host Settings
+  blockinfile:
+    state: present
+    block: |
+      $settings['trusted_host_patterns'] = array(
+      {% for host in drupal_trusted_hosts %}
+        '{{ host }}',
+      {% endfor %}
+      );
+    path: "{{ drupal_trusted_hosts_file }}"
+    marker: // {mark} ANSIBLE MANAGED BLOCK - trusted hosts
+  when: webserver_app_do_trusted_host
+
+- name: Flysystem "fedora" scheme configuration
+  blockinfile:
+    state: present
+    block: |
+      $settings['flysystem'] = [
+        'fedora' => [
+          'driver' => 'fedora',
+          'config' => [
+            'root' => '{{ fedora_base_url }}',
+          ],
+        ],
+      ];
+    path: "{{ drupal_site_default_settings }}"
+    marker: // {mark} ANSIBLE MANAGED BLOCK - fedora flysystem scheme
+  when: webserver_app_do_fedora_scheme_config
+
+- name: Content Sync Directory configuration
+  blockinfile:
+    state: present
+    block: |
+      global $content_directories;
+      $content_directories['sync'] = $app_root.'/../content/sync';
+    path: "{{ drupal_site_default_settings }}"
+    marker: // {mark} ANSIBLE MANAGED BLOCK - content sync dir
+  when: webserver_app_do_content_sync_config
```
Not sure what the best approach was here, to allow for more granular control over what was being done in what was originally called the "Trusted Host Settings" bit, which was doing more than just setting the "trusted hosts" settings... elected to keep the original, but additionally have all the bits it contained be available in separate chunks, as we don't really want the `content/sync` thing being done in the `starter` "profile".

adam-vessey commented on roles/internal/webserver-app/tasks/drupal.yml:
```diff
@@ -48,24 +90,28 @@
     path: "{{ drupal_deploy_dir }}/content/sync"
     owner: "{{ webserver_app_user }}"
     group: "{{ webserver_app_user }}"
+  when: webserver_app_do_content_sync_config or webserver_app_do_original_trusted_host_and_more
 
 - name: Import features
   command: "{{ drush_path }} --root {{ drupal_core_path }} -y fim islandora_core_feature,islandora_defaults"
-    
+  when: webserver_app_feature_import_base
+
 - name: Import search feature
   command: "{{ drush_path }} --root {{ drupal_core_path }} -y fim islandora_search"
-  when: islandora_profile == "standard"
+  when: webserver_app_feature_import_search
 
 # Set Matomo Settings
 - name: Set Matomo site id.
   command: "{{ drush_path }} --root {{ drupal_core_path }} -y config-set matomo.settings site_id 1"
-  register: set_search_api_config
-  changed_when: "'Do you want to update site_id' in set_search_api_config.stdout"
+  register: set_matomo_site_id_config
+  changed_when: "'Do you want to update site_id' in set_matomo_site_id_config.stdout"
+  when: webserver_app_configure_matomo
 
-- name: Set Matmo sever url.
+- name: Set Matomo server url.
   command: "{{ drush_path }} --root {{ drupal_core_path }} -y config-set matomo.settings url_http http://localhost:8000/matomo/"
-  register: set_search_api_config
-  changed_when: "'Do you want to update site_id' in set_search_api_config.stdout"
+  register: set_matomo_server_url_config
+  changed_when: "'Do you want to update site_id' in set_matomo_server_url_config.stdout"
+  when: webserver_app_configure_matomo
```
The `register`/`changed_when` things touching `set_search_api_config` seems like something of copypasta error, as we're not dealing with "search API" here... so adjusted things in a way that seems more consistent?

... it's... _possible_ that the second will have to have the string it's searching for changed... instead of `site_id`, to be `url_http` or something of the like.

adam-vessey replied:
> ... it's... _possible_ that the second will have to have the string it's searching for changed... instead of `site_id`, to be `url_http` or something of the like.

Did indeed appear to be the case. Running the command manually: 
![Screen Shot 2022-09-13 at 12 10 52 PM](https://user-images.githubusercontent.com/607975/189938819-25d2a96a-30eb-4b60-836c-efb47b488cae.png)

Addressed in https://github.com/Islandora-Devops/islandora-playbook/pull/226/commits/dd550679f02cef9a3bc1961f2efc97d1981033fe

adam-vessey commented on vars/starter.yml:
```diff
@@ -0,0 +1,24 @@
+---
+drupal_build_composer_project: true
+drupal_composer_dependencies: []
+# Presently pointing at the main branch.
+drupal_composer_project_package: 'islandora/islandora-starter-site:dev-main'
```
This should eventually be pointed at whatever reference structure we move to use for releases... or maybe left without a version reference to pull whatever "latest" version?

adam-vessey commented on vars/starter_dev.yml:
```diff
@@ -0,0 +1,34 @@
+---
+drupal_build_composer_project: false
+drupal_deploy: true
+# XXX: Just pointing at my repo for the moment, for testing... should be pointed
+# back at Islandora's prior to merging.
+drupal_deploy_repo: https://github.com/adam-vessey/islandora-starter-site.git
+# XXX: Should be made to point at "main".
+drupal_deploy_version: implementation
```
Yeah, needs to be pointed at the merged bits, whenever it's possible to do so.

adam-vessey replied:
Done in https://github.com/Islandora-Devops/islandora-playbook/pull/226/commits/a18bde06eaa5a4b0e54c3b87b6e400a716184c1d

adam-vessey commented on vars/starter.yml:
```diff
@@ -0,0 +1,24 @@
+---
+drupal_build_composer_project: true
+drupal_composer_dependencies: []
```
Strictly, doesn't need to be specified here (the default in the `geerlingguy.drupal` is an empty list); however, thought to leave it to maintain something of a parallel structure with the `demo` and `standard` "profiles"... could be useful for people looking to hack on things?

adam-vessey replied:
Adjusted, to leave with a comment and with it commented out.

adam-vessey commented on vars/starter.yml:
```diff
@@ -0,0 +1,24 @@
+drupal_install_profile: minimal
+
+# None necessary, due to using an existing configuration.
+drupal_enable_module: []
```
Same deal: Strictly, doesn't need to be specified here (the default in the `geerlingguy.drupal` is an empty list); however, thought to leave it to maintain something of a parallel structure with the `demo` and `standard` "profiles"... could be useful for people looking to hack on things?

adam-vessey replied:
Adjusted, to leave with a comment and with it commented out.

adam-vessey committed:
Fix use of `drupal_enable_modules` variable in starter profiles.

adam-vessey committed:
Comment out the things being kept around for parallel structures.

adam-vessey commented on roles/internal/webserver-app/tasks/solr.yml:
```diff
@@ -6,12 +6,12 @@
   register: set_search_api_config_host
-  changed_when: "'Do you want to update' in set_search_api_config.stdout"
+  changed_when: "'Do you want to update' in set_search_api_config_host.stdout"
 
 - name: Set default solr server to point to first core
   command: "{{ drush_path }} --root {{ drupal_core_path }} -y config-set search_api.server.default_solr_server backend_config.connector_config.core {{ solr_cores[0] }}"
   register: set_search_api_config_core
-  changed_when: "'Do you want to update' in set_search_api_config.stdout"
+  changed_when: "'Do you want to update' in set_search_api_config_core.stdout"
```
Looking back at https://github.com/Islandora-Devops/islandora-playbook/pull/126/files#diff-5bddb401e18178b7584b51cd9831eaa348385079b2224a85e3db8fa2412fc547L5-R15 ... I believe this is the intent, as otherwise, it seems that the `set_search_api_config` bit would've been being registered by the "Matomo" setup, in the `webserver-app/tasks/drupal.yml` bit?

rosiel replied:
Yeah that seems like a typo in the Matomo stuff.

adam-vessey committed:
Point at the merged starter site.

rosiel commented on vars/starter.yml:
```diff
@@ -0,0 +1,27 @@
+drupal_site_install_extra_args:
+  - '--existing-config'
+
+openseadragon_composer_require: false
```
Some of these - setting up Matomo and FITS - seem like they're `false` only because theyre out of scope at the moment but we probably eventually want them. Can you denote/comment/group them in some way to distinguish from the others which we do/don't need due to the architecture of the Starter Site?

adam-vessey replied:
Took a shot in https://github.com/Islandora-Devops/islandora-playbook/pull/226/commits/52f962aa6aad7901370e2d634c40991c51efe198 ... thoughts?

adam-vessey replied:
I feel like I should mention: Anything that _does_ need to do any configuration probably maybe shouldn't do it by `config:set`'ing anything, but instead to use other mechanisms such as `config_override` to avoid potentially dirtying configs... to avoid specific environmental concerns creeping into things; for example: If the `matomo.settings:url_http` value was made to dynamically point at the hostname instead of just specifying `localhost:8000`, then yeah, it wouldn't really belong in the config that gets used between environments where that remote would(/could) vary.

... This... _does_ kind of get into a weird area with [how the JWT stuff is presently configured](https://github.com/Islandora-Devops/islandora-playbook/blob/311aa2126947869ce20d7765e9689de3d219d835/roles/internal/webserver-app/tasks/jwt.yml#L10-L39), in that because things were based on the "standard" playbook install, that it just so happens to match the paths; however, if other paths are ever specified for `webserver_app_jwt_key_path`, then it could end up diverging...

adam-vessey committed:
Group variables based to some degree on function.

adam-vessey committed:
Update the target string accordingly.

bibliophileaxe commented on roles/internal/webserver-app/tasks/drupal.yml:
```diff
@@ -1,6 +1,6 @@
 ---
 
-- name: Trusted Host Settings
+- name: Trusted Host Settings and more
```
Maybe we can rename this to settings.php settings or something generic?

adam-vessey replied:
There's more than just this singular task doing `settings.php` things... really, was just trying to reflect the reality in the name that it was doing more than just the "trusted host" stuff it was originally named... Like the flysystem and `$content_directories` stuff... and trying to follow something of the established patterns of naming things.

adam-vessey committed:
Fix composer invocation to _not_ run as root.

Seems like Composer 2.4.2
(https://github.com/composer/composer/releases/tag/2.4.2) became more
strict, preventing the command from returning completely.

adam-vessey committed:
Merge remote-tracking branch 'origin/dev' into feature/starter-site

adam-vessey committed:
Back to doing the Matomo and FITS things.

adam-vessey committed:
Since we now do Matomo and FITS, can drop the notes indicating we didn't.

adam-vessey commented on vars/starter_dev.yml:
```diff
@@ -0,0 +1,32 @@
+---
+drupal_build_composer_project: false
+drupal_deploy: true
+#drupal_deploy_repo: https://github.com/Islandora/islandora-starter-site.git
+#drupal_deploy_version: main
+drupal_deploy_repo: https://github.com/adam-vessey/islandora-starter-site.git
+drupal_deploy_version: feature/playbook-standard-replacement
```
Back to needing to be pointed back upstream after https://github.com/Islandora/islandora-starter-site/pull/6 is merged.

adam-vessey replied:
Addressed in https://github.com/Islandora-Devops/islandora-playbook/pull/226/commits/b3f1bca7b8f8dfe18f16b10e60e259e3a984395e

adam-vessey committed:
Fix another instance of composer running as root.

adam-vessey committed:
Merge remote-tracking branch 'origin/dev' into feature/starter-site

adam-vessey committed:
Add in the other migration.

Also, allow the attempted chown'ing of islandora_defaults to be
skipped... 'cause the starter stuff doesn't/shouldn't have it.

adam-vessey committed:
Drop `standard` "profile"

adam-vessey committed:
Point back upstream for the referenced repo.

adam-vessey committed:
Include the dev things in the dev thing.
